### ISSUE TITLE
Strimzi operator changing labels that Flux applies causing resource recreation

### ISSUE THREAD

**Issue by minesskylineGTR** [fluxcd/kustomize-controller#296]
This is a continuation of a thread/discussion with @stefanprodan about Flux2 recreating resources for Kafka that have no manual changes taking place from Slack (https://cloud-native.slack.com/archives/CLAJ40HV3/p1612317051175100). He asked me to provide a repo to recreate the issue I am experiencing, which can be found here:
https://github.com/minesskylineGTR/flux2-kafka

I have added Stefan as a Collaborator to the repo, but please let me know if anybody else needs access. I have the latest logs from the kustomize-controller as requested:

`{"level":"info","ts":"2021-03-09T03:17:24.744Z","logger":"controllers.Kustomization","msg":"Reconciliation finished in 1.783939721s, next run in 10m0s","controller":"kustomization","request":"flux-system/apps-dev","revision":"master/2d6eaf1c823ac16f114a1ec30ca87fbe017b9a9f"}
{"level":"info","ts":"2021-03-09T03:18:40.655Z","logger":"controllers.Kustomization","msg":"Kustomization applied in 419.847329ms","kustomization":"flux-system/infrastructure-dev","output":{"clusterrole.rbac.authorization.k8s.io/cluster-autoscaler":"unchanged","clusterrolebinding.rbac.authorization.k8s.io/cluster-autoscaler":"unchanged","deployment.apps/cluster-autoscaler":"unchanged","helmrelease.helm.toolkit.fluxcd.io/jaeger-operator":"unchanged","helmrelease.helm.toolkit.fluxcd.io/kafka-operator":"unchanged","helmrelease.helm.toolkit.fluxcd.io/nginx-ingress":"unchanged","helmrepository.source.toolkit.fluxcd.io/helm-charts-stable":"unchanged","helmrepository.source.toolkit.fluxcd.io/influxdb2":"unchanged","helmrepository.source.toolkit.fluxcd.io/jaeger-operator":"unchanged","helmrepository.source.toolkit.fluxcd.io/kube-prometheus-stack":"unchanged","helmrepository.source.toolkit.fluxcd.io/nginx-ingress":"unchanged","helmrepository.source.toolkit.fluxcd.io/strimzi-kafka-operator":"unchanged","namespace/jaeger":"unchanged","namespace/kafka":"unchanged","namespace/nginx-ingress":"unchanged","role.rbac.authorization.k8s.io/cluster-autoscaler":"unchanged","rolebinding.rbac.authorization.k8s.io/cluster-autoscaler":"unchanged","secret/default-server-secret":"unchanged","serviceaccount/cluster-autoscaler":"unchanged","storageclass.storage.k8s.io/encrypted-gp2":"unchanged"}}
{"level":"info","ts":"2021-03-09T03:18:40.670Z","logger":"controllers.Kustomization","msg":"Reconciliation finished in 1.291520772s, next run in 10m0s","controller":"kustomization","request":"flux-system/infrastructure-dev","revision":"master/2d6eaf1c823ac16f114a1ec30ca87fbe017b9a9f"}
{"level":"info","ts":"2021-03-09T03:18:45.900Z","logger":"controllers.Kustomization","msg":"All dependencies area ready, proceeding with reconciliation","controller":"kustomization","request":"flux-system/kafka-dev"}
{"level":"info","ts":"2021-03-09T03:18:46.702Z","logger":"controllers.Kustomization","msg":"Kustomization applied in 414.946253ms","kustomization":"flux-system/kafka-dev","output":{"configmap/kafka-metrics":"configured","configmap/strimzi-cluster-operator":"configured","kafka.kafka.strimzi.io/kafka":"configured","kafkarebalance.kafka.strimzi.io/kafka-rebalance":"configured","kafkatopic.kafka.strimzi.io/telegraf":"configured","kafkauser.kafka.strimzi.io/telegraf":"configured"}}
{"level":"info","ts":"2021-03-09T03:18:46.719Z","logger":"controllers.Kustomization","msg":"Reconciliation finished in 834.466143ms, next run in 10m0s","controller":"kustomization","request":"flux-system/kafka-dev","revision":"master/2d6eaf1c823ac16f114a1ec30ca87fbe017b9a9f"}
{"level":"info","ts":"2021-03-09T03:21:24.970Z","logger":"controllers.Kustomization","msg":"Kustomization applied in 702.201235ms","kustomization":"flux-system/flux-system","output":{"clusterrolebinding.rbac.authorization.k8s.io/cluster-reconciler-flux-system":"unchanged","customresourcedefinition.apiextensions.k8s.io/alerts.notification.toolkit.fluxcd.io":"configured","customresourcedefinition.apiextensions.k8s.io/buckets.source.toolkit.fluxcd.io":"configured","customresourcedefinition.apiextensions.k8s.io/gitrepositories.source.toolkit.fluxcd.io":"configured","customresourcedefinition.apiextensions.k8s.io/helmcharts.source.toolkit.fluxcd.io":"configured","customresourcedefinition.apiextensions.k8s.io/helmreleases.helm.toolkit.fluxcd.io":"configured","customresourcedefinition.apiextensions.k8s.io/helmrepositories.source.toolkit.fluxcd.io":"configured","customresourcedefinition.apiextensions.k8s.io/kustomizations.kustomize.toolkit.fluxcd.io":"configured","customresourcedefinition.apiextensions.k8s.io/providers.notification.toolkit.fluxcd.io":"configured","customresourcedefinition.apiextensions.k8s.io/receivers.notification.toolkit.fluxcd.io":"configured","deployment.apps/helm-controller":"configured","deployment.apps/kustomize-controller":"configured","deployment.apps/notification-controller":"configured","deployment.apps/source-controller":"configured","gitrepository.source.toolkit.fluxcd.io/flux-system":"unchanged","kustomization.kustomize.toolkit.fluxcd.io/apps-dev":"unchanged","kustomization.kustomize.toolkit.fluxcd.io/flux-system":"unchanged","kustomization.kustomize.toolkit.fluxcd.io/infrastructure-dev":"unchanged","kustomization.kustomize.toolkit.fluxcd.io/kafka-dev":"unchanged","namespace/flux-system":"unchanged","networkpolicy.networking.k8s.io/allow-scraping":"unchanged","networkpolicy.networking.k8s.io/allow-webhooks":"unchanged","networkpolicy.networking.k8s.io/deny-ingress":"unchanged","role.rbac.authorization.k8s.io/crd-controller-flux-system":"unchanged","rolebinding.rbac.authorization.k8s.io/crd-controller-flux-system":"unchanged","service/notification-controller":"unchanged","service/source-controller":"unchanged","service/webhook-receiver":"unchanged"}}
{"level":"info","ts":"2021-03-09T03:21:24.991Z","logger":"controllers.Kustomization","msg":"Reconciliation finished in 2.060434874s, next run in 10m0s","controller":"kustomization","request":"flux-system/flux-system","revision":"master/2d6eaf1c823ac16f114a1ec30ca87fbe017b9a9f"}
{"level":"info","ts":"2021-03-09T03:27:24.762Z","logger":"controllers.Kustomization","msg":"All dependencies area ready, proceeding with reconciliation","controller":"kustomization","request":"flux-system/apps-dev"}
{"level":"info","ts":"2021-03-09T03:27:26.495Z","logger":"controllers.Kustomization","msg":"Kustomization applied in 567.365137ms","kustomization":"flux-system/apps-dev","output":{"clusterrole.rbac.authorization.k8s.io/cloudwatch-agent-role":"unchanged","clusterrole.rbac.authorization.k8s.io/default-psp-monitoring-dev":"unchanged","clusterrole.rbac.authorization.k8s.io/fluentd-role":"unchanged","clusterrolebinding.rbac.authorization.k8s.io/cloudwatch-agent-role-binding":"unchanged","clusterrolebinding.rbac.authorization.k8s.io/fluentd-role-binding":"unchanged","configmap/cluster-info":"unchanged","configmap/cwagentconfig":"unchanged","configmap/fluentd-config":"unchanged","configmap/kafka-metrics":"configured","configmap/strimzi-cluster-operator":"configured","daemonset.apps/cloudwatch-agent":"unchanged","daemonset.apps/fluentd-cloudwatch":"unchanged","helmrelease.helm.toolkit.fluxcd.io/gitlab-runner":"unchanged","helmrelease.helm.toolkit.fluxcd.io/influxdb2":"unchanged","helmrelease.helm.toolkit.fluxcd.io/prom-operator":"unchanged","jaeger.jaegertracing.io/jaeger":"configured","kafka.kafka.strimzi.io/kafka":"configured","kafkarebalance.kafka.strimzi.io/kafka-rebalance":"configured","kafkatopic.kafka.strimzi.io/telegraf":"configured","kafkauser.kafka.strimzi.io/telegraf":"configured","namespace/amazon-cloudwatch":"unchanged","namespace/gitlab-runner":"unchanged","namespace/monitoring-dev":"unchanged","podsecuritypolicy.policy/default-psp-monitoring-dev":"configured","rolebinding.rbac.authorization.k8s.io/default-psp-monitoring-dev":"unchanged","secret/dev-tls-certificate":"unchanged","serviceaccount/cloudwatch-agent":"unchanged","serviceaccount/fluentd":"unchanged"}}
{"level":"info","ts":"2021-03-09T03:27:26.511Z","logger":"controllers.Kustomization","msg":"Reconciliation finished in 1.763796998s, next run in 10m0s","controller":"kustomization","request":"flux-system/apps-dev","revision":"master/2d6eaf1c823ac16f114a1ec30ca87fbe017b9a9f"}`

My `flux check` output:

`flux check`
► checking prerequisites
✔ kubectl 1.20.4 >=1.18.0-0
✔ Kubernetes 1.18.9-eks-d1db3c >=1.16.0-0
► checking controllers
✔ helm-controller: healthy
► ghcr.io/fluxcd/helm-controller:v0.2.1
✔ kustomize-controller: healthy
► ghcr.io/fluxcd/kustomize-controller:v0.2.2
✔ notification-controller: healthy
► ghcr.io/fluxcd/notification-controller:v0.2.1
✔ source-controller: healthy
► ghcr.io/fluxcd/source-controller:v0.2.2
✔ all checks passed

Flux version:

`flux -v`

`flux version 0.9.1`

If there is anything that I missed, please let me know. There is a better than not chance I have done something wrong, or have a misunderstanding of how Flux is designed to work.

**Comment by stefanprodan** [comment]
@minesskylineGTR please upgrade you controllers to the latest version, you are running a very old control plane

**Comment by minesskylineGTR** [comment]
@stefanprodan I found this post which is the same exact issue I am experiencing. I've done the same kustomization change as this post, but with the addition of suspending the kustomization from the Flux side:

https://www.gitmemory.com/issue/fluxcd/flux2/414/721077327

**Comment by stefanprodan** [comment]
The checksum label is set only when pruning is enabled, set prune: false. I would reach out to the Kafka controller people and ask for a feature to ignore certain labels or disable labels comparison.

**Comment by minesskylineGTR** [comment]
@stefanprodan Thanks for that. I do have `prune: false` set already:

```apiVersion: kustomize.toolkit.fluxcd.io/v1beta1
kind: Kustomization
metadata:
  name: kafka-dev
  namespace: flux-system
spec:
  interval: 10m0s
  dependsOn:
    - name: infrastructure-dev
  sourceRef:
    kind: GitRepository
    name: flux-system
  path: ./us-west-2/kafka
  prune: false
  validation: client```

I will reach out to the Strimzi maintainers and see what they recommend for setting static labels.

**Comment by stefanprodan** [comment]
So if no labels value are changed by Flux why does the Kafka controller recreates objects?

PS: Please change the title as it’s misleading, it’s not Flux that recreates but some other operator.

**Comment by bigkevmcd** [comment]
I suspect that it's caused by this issue https://github.com/strimzi/strimzi-kafka-operator/issues/4394

**Comment by minesskylineGTR** [comment]
> So if no labels value are changed by Flux why does the Kafka controller recreates objects?
> 
> PS: Please change the title as it’s misleading, it’s not Flux that recreates objects but some other operator.

I've changed the title to better reflect. I am not sure why the operator feels the need to change labels for those resources. I had assumed it was Flux because it was happening every 10 minutes without fail, which is the interval time on the kustomization that's set to check. I did get some further information from the Strimzi maintainer to set static labels, so I will go that route and close this. Thanks for taking a look!
